src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
+            auto server_session = static_cast<ServerSession*>(object);
+            auto result = server_session->ReceiveIPCRequest(this);
+            SyncOutput(result, idx);
+        } else {
+            SyncOutput(RESULT_SUCCESS, idx);
+        }
+        break;
+    }
 
-        auto& vm_manager = owner_process->vm_manager;
+    case Status::WaitHleEvent:
+        object->Acquire(this);
+        hle_context->OnWakeUp(this, WakeupReason::Signal, hle_callback);
+        hle_context = nullptr;
+        hle_callback = nullptr;
 
-        // Map the page to the current process' address space.
-        vm_manager.MapBackingMemory(Memory::TLS_AREA_VADDR + available_page * Memory::PAGE_SIZE,
-                                    memory.GetFCRAMRef(*offset), Memory::PAGE_SIZE,
-                                    MemoryState::Locked);
+    default:
+        ASSERT_MSG(false, "Inconsistent thread status");
     }
 
-    // Mark the slot as used
-    tls_slots[available_page].set(available_slot);
-    thread->tls_address = Memory::TLS_AREA_VADDR + available_page * Memory::PAGE_SIZE +
-                          available_slot * Memory::TLS_ENTRY_SIZE;
+    // Clear waiting objects/threads
+    for (auto obj : waiting_on) {
+        obj->RemoveWaitingThread(this);
+    }
+    waiting_on.clear();
+    core.root.UnscheduleEvent(wakeup_event.get());
+    status = Thread::Status::Ready;
+}
 
-    memory.ZeroBlock(*owner_process, thread->tls_address, Memory::TLS_ENTRY_SIZE);
+void Thread::SetPriority(u32 priority) {
+    nominal_priority = priority;
+    for (auto obj : waiting_on) {
+        auto mutex = dynamic_cast<Kernel::Mutex*>(obj.get());
+        if (mutex && mutex->holding_thread) {
+            mutex->holding_thread->real_priority.reset();
+        }
+    }
+}
+
+void Thread::WakeUp() {
+    switch (status) {
+    case Status::WaitSyncAny:
+        SyncOutput(RESULT_TIMEOUT, -1);
+        break;
 
-    // TODO(peachum): move to ScheduleThread() when scheduler is added so selected core is used
-    // to initialize the context
-    ResetThreadContext(thread->context, stack_top, entry_point, arg);
+    case Status::WaitHleEvent:
+        hle_context->OnWakeUp(this, WakeupReason::Timeout, hle_callback);
+        hle_context = nullptr;
+        hle_callback = nullptr;
+        break;
 
-    thread_managers[processor_id]->ready_queue.push_back(thread->current_priority, thread.get());
-    thread->status = ThreadStatus::Ready;
+    case Status::WaitSyncAll:
+        SyncOutput(RESULT_TIMEOUT);
+        break;
 
-    return MakeResult<std::shared_ptr<Thread>>(std::move(thread));
+    case Status::WaitSleep:
+        break;
+
+    case Status::WaitArb:
+        arbiter->RemoveWaitingThread(this);
+        break;
+
+    default:
+        ASSERT_MSG(false, "Inconsistent thread status");
+    }
+
+    // Clear waiting objects/threads
+    for (auto obj : waiting_on) {
```

Comment:
```suggestion
    for (auto& obj : waiting_on) {
```

otherwise this can cause atomic reference count increment/decrement churn.